### DISCUSSION TITLE
Instructor Led Training (ILT) Feature has been developed by Protean eGov Technologies Limited and wishes to contribute it to Community

### DISCUSSION THREAD

**Discussion by sujitrs** [Sunbird-Ed/Community#4] [Contribution]
Instructor Led Training (ILT) Feature has been developed by Protean eGov Technologies Limited and wishes to contribute it to Community. Please suggest / guide with next steps.
@surendrasinghs

**Comment by pramodkvarma** [comment]
That's wonderful @sujitrs! Can you please share a document/wiki/URL that points to the features, UX, and its tech designs? I would like the Sunbird-ED community to read and comment.

**Comment by sujitrs** [comment]
Please refer below documents for Features and UX:
[ILT Flow Mobile.pdf](https://github.com/Sunbird-Ed/Community/files/8553401/ILT.Flow.Mobile.pdf)
[ILT Flow Web.pdf](https://github.com/Sunbird-Ed/Community/files/8553402/ILT.Flow.Web.pdf)

**Reply by pramodkvarma** [reply]
This is excellent! Event and synchronous sessions have been in the list of features for a while. Wonderful to see it built by the community and being contributed back.

@krishnaj20 @kameshbhr @alok-os @jagadishbabu please check the feature design and discuss here.

@sujitrs can you also please attach tech design document (which APIs are leveraged, any changes to the core, deployment diagram, etc) along with the URL to the codebase 

@vrayulu requesting your comments and feedback as well.

Tagging @parthlawate as well

**Reply by kameshbhr** [reply]
Yes, this will be an excellent add on to the Sunbird capabilities. Here are quick comments and queries:

1. Event types - Instead of naming "Classroom Event" "Virtual Event" - calling them "Offline" and "Online" events may be more intuitive (at least internally).
2. Users not being able join the event until the moderator joins - This is shown as part of the workflow. I am not sure if this would be possible, or needed to be there within Sunbird. Once users try to join, they will be redirected to the online provider app. That will handle it anyway.
3. The workflow of "joining" an Offline event is not clear in the current flow. If it is an offline event, it would be an overhead for the user to just open the app and click "Join". 
4. Attendance dashboard - Where will the information of Join and Leave Time come from? From the online provider app? All of them may not provide it. For offline events, it has to be anyway given by users. If it is user driven, it wouldn't be accurate.
5. We should use a different role other than "Content Creator" for event creation. We can have a specific "Event Creator" role. 
6. It may be better to have event creation within Events page, instead of Workspace. Do events require review workflow?
7. We should use common terminology either "Enroll" or "Register" everywhere.  

Would like to go through the tech docs as well.

**Reply by pramodkvarma** [reply]
Thanks @kameshbhr for quick comments. I will let @sujitrs respond as required. 

One quick comment from my side. Some of the items you mentioned (event type naming, button naming, to enable user to click join even for offline events, etc) should be anyway configuration (that's idea of generalized building block). Means, we don't have to be opinionated as long as they are configurable. 

Let us make sure code, design, configurability, telemetry, deployment, etc are done as per typical Sunbird building block and then we can bring the contribution in. Makes sense, right?

**Reply by vrayulu** [reply]
Thanks @sujitrs for the contribution. Please share the tech/design docs for this feature. And are you aware of the changes made by iGOT team (another Sunbird contributor) for events capability in Sunbird and built this on top of them? if yes, thats awesome. else, lets align both the changes and plan for merging back into the core.

**Reply by parthlawate** [reply]
This is great news ! Thanks @sujitrs & Protean !! This will be really appreciated by the community ...

@vrayulu yes the solution is expanded on top of what the iGot team did. The Online Provider integrations are done as an Adaptor so we can add more for newer providers. 

Adding Ashwin Date @coolbung to give more context.

**Reply by pramodkvarma** [reply]
Fantastic! I also saw OSS online conferencing solution integration as an option. Cool!

**Reply by parthlawate** [reply]
yes right now its Big Blue Button. But we can add other adaptors

**Reply by pramodkvarma** [reply]
Awesome! Do share design docs here. Thanks.

**Reply by alok-os** [reply]
excellent !!! Couple of suggestions and observations:
2. We should define or use an existing spec (if it exist) for the interface definition for sync session service providers - keeping the interface to minimum. The document seem to reflect much tighter coupling - was wondering is that the best idea. For sync sessions - basic info. there is always a possibility for the user to join directly without going thru event - we need to keep that in mind. ex The document seem to reflect that App will wait if moderator has not joined - do we need that? Do we want to enable storing session recordings in sunbird App by user? 
3. Creation thru mobile - interesting idea. we have enabled creation only on portal app - but make sense to enable it on App for events. let us bring it in a way so that other type of asset creation can be added later if reqd with minimal effort
4. Offline flow was not very clear - will be good to validate the flows with real use cases.

**Reply by luckynvdu** [reply]
Hey @sujitrs , This is really awesome to have such contribution. We could help you to get this integrated on mobile and portal as well.

We have a few open questions -
1. What is the criteria of discoverablility of events for users. E.x. If I belong to a specific framework then is there any way to taregt me or will be generic targeting to all
2. What could be the probable filters for events and are they configurable? E.x. If we can choose to add/remove filters and their values
3. Could you help us with the tech stack information as well?
@sharathkashyap

**Reply by pramodkvarma** [reply]
@alok-os this one is for offering it as native capability. If the entire service is plugged in through 3rd party then what you are saying is right. Here the capability is inbuilt and only video/audio conferencing is a plugin.

**Reply by coolbung** [reply]
@baljeetsingh is sharing a note about the current status of the feature from a Product and Tech standpoint (although Sujit's PDFs do already cover a lot of the product side). 

The next step I think should be to 
1. Identify what would be the scope for MVP
2. Identify gaps between MVP and current implementation
3. Contribute the feature to Sunbird 5.0/5.1

Do let me know if this looks like a reasonable way forward

**Reply by pramodkvarma** [reply]
I would prefer to bring the already developed feature in (even if not optimal UX, as long as code and structure is OK) and then iterate to improve rather than perfecting it before contribution. This is important us to not go into a "perfect design trap" @coolbung @alok-os @kameshbhr @luckynvdu @sharathkashyap @vrayulu 

@coolbung @baljeetsingh please share design docs, APIs, and pointer to code. Thanks.

**Reply by parthlawate** [reply]
@pramodkvarma Agree. Once we have more adoption, we will know what improvements to prioritise. Basically consider what is there right now as an MVP

**Reply by Krishnaj20** [reply]
looping in @rhwarrier

**Comment by baljeetsingh** [comment]
###  **Overview about frontend / backend** 

Contribution initiated by **Protean eGov Technologies Ltd** (Formerly NSDL e-Governance Infrastructure Limited )
For more details / information please reach out to Protean team.

**Reply by baljeetsingh** [reply]
**Frontend**
Event Library is the Angular NPM library.
This library is created for view, create and get a list of events and also it handles the time zones of event time.

In this library we have created separate components and services therefore users can use UI and functionality independently as he/she requires.

The best part of the libraries, the implementer can use UI and services separately for better compatibility with any site.

**Views in the module:**

**Event list -** This is a view where we can list down all the events. In this view, we reused the common consumption library for the card grid.

**Event form -** this is a view where we can generate new events. or edit old events. We can save as a draft and also can publish the event via form. We implemented this with sb-form and asset library integration

**Event Detail -** Here We can see a detailed view of the event. In this component we implemented 3 separate components that are- cover component, buttons component, advance detail component to make is more dynamic and reusable

We can See event recordings also in this page (BBB event provider is integrated, recording needs to be started by moderator)

**Event Dashboard -** This view can be seen after clicking the view dashboard button on the detail page. This is only accessible to the creator. In this, we can see general attendance as well as detailed user attendance for every attendee

**Event Calendar -** This is a view where we can see all events on a calendar view according to day, week, month tabs. Also we have separate list view on this page which shows events of particular date in card format

**Event filters -** This is a customized filter component developed with the help of sb-form and can be integrated with List view and calendar view as well.

Documentation for integration: [Integration document](https://docs.google.com/document/d/1bsN_RXKBfLwV457PCL0gVeyMk-n1NpulOEJYTeMDtos/edit#)
Library Repo: [Event library](
https://github.com/NSDL-e-Governance/protean-egov-sb-events-module/tree/private-repo-branch)
NPM link: [ngtek-event-library](https://www.npmjs.com/package/ngtek-event-library)

**Reply by baljeetsingh** [reply]
**Backend**

In the Knowledge-Platform Content-service release-4.2.0,  over the basic Event CRUD apis, we have integrated the online provider support for the Events. 

This online provider integration is implemented generic using interfaces and completed with the BigBlueButton(BBB) integration, where we can create meetings, get the meeting links to join for both moderator as well as attendee.

In sunbird-course-service release-4.2.0, we have ported the code of Nitin Puri(Univermal) to support the Event and EventSet for batches and enrollment- https://github.com/project-sunbird/sunbird-course-service/pull/331. 

In the course service, we have added generic apis for creating, getting event attendance and getting the recordings of the event for now with BBB integration using webhooks.

**We have integrated above in 2 services :**

1.  **Knowledge-Platform : Content-Api**
       Repository link : [knowledge-platform (release-4.2.0)](https://github.com/NSDL-e-Governance/protean-egov-knowledge-platform/tree/release-4.2.0_nsdl)

2.  **Sunbird-Course-Service :**
       Repository link : [sunbird-course-service](https://github.com/NSDL-e-Governance/protean-egov-sunbird-course-service/tree/release-4.2.0_nsdl)
       
Link for the detailed document - [Backend Implementation done](https://docs.google.com/document/d/18COYG6a5LmW2bFHx7u1X9P_eHSSAJnESwU1eKVyplUg/edit#heading=h.ft24o1tcvzm7)

**Comment by luckynvdu** [comment]
@baljeetsingh @parthlawate could you please create an RFC for the same and create an epic for the same in the new issue tracker as well

**Reply by parthlawate** [reply]
Done

**Reply by chitranshu-keshav** [reply]
@parthlawate can you please add the link of the EPIC/RFC in this thread

**Reply by parthlawate** [reply]
here you go.. Note that right now its just linked to this thread. Seeking some clarity/templates to follow so the JIRA Epic & related items can be populated 

https://project-sunbird.atlassian.net/browse/ED-236
